c/client.c.go

```diff
@@ -0,0 +1,223 @@
+	}
+
+	if transitRelayUrl != nil {
+		client.TransitRelayURL = C.GoString(transitRelayUrl)
+	}
+
+	if passPhraseComponentLength == 0 {
+		client.PassPhraseComponentLength = int(passPhraseComponentLength)
+	}
+
+	clientPtr := uintptr(unsafe.Pointer(client))
+	clientsMap[clientPtr] = client
+
+	return clientPtr
+}
+
+//export FreeClient
+func FreeClient(clientPtr uintptr) int {
+	if _, err := getClient(clientPtr); err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+
+	delete(clientsMap, clientPtr)
+	return int(codes.OK)
+}
+
+//export ClientSendText
+func ClientSendText(ctxC unsafe.Pointer, clientPtr uintptr, msgC *C.char, codeOutC **C.char, cb C.callback) int {
+	client, err := getClient(clientPtr)
+	if err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+	ctx := context.Background()
+
+	code, status, err := client.SendText(ctx, C.GoString(msgC))
+	if err != nil {
+		log.Printf("%v\n", err)
+		return int(codes.ERR_SEND_TEXT)
+	}
+	fmt.Printf("code returned: %s\n", code)
+	*codeOutC = C.CString(code)
+
+	go func() {
+		s := <-status
+		if s.Error != nil {
+			// TODO: stick error message somewhere conventional for C to read.
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_SEND_TEXT_RESULT))
+		} else if s.OK {
+			C.call_callback(ctxC, cb, nil, C.int(codes.OK))
+		} else {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_UNKNOWN))
+		}
+	}()
+
+	return int(codes.OK)
+}
+
+//export ClientSendFile
+func ClientSendFile(ctxC unsafe.Pointer, clientPtr uintptr, fileName *C.char, length C.int, fileBytes unsafe.Pointer, codeOutC **C.char, cb C.callback) int {
+	client, err := getClient(clientPtr)
+	if err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+	ctx := context.Background()
+
+	// TODO: is there a way to avoid copying?
+	reader := bytes.NewReader(C.GoBytes(fileBytes, length))
+
+	code, status, err := client.SendFile(ctx, C.GoString(fileName), reader, false)
+	if err != nil {
+		return int(codes.ERR_SEND_TEXT)
+	}
+	*codeOutC = C.CString(code)
+
+	go func() {
+		s := <-status
+		if s.Error != nil {
+			// TODO: stick error message somewhere conventional for C to read.
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_SEND_TEXT_RESULT))
+		} else if s.OK {
+			C.call_callback(ctxC, cb, nil, C.int(codes.OK))
+		} else {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_UNKNOWN))
+		}
+	}()
+
+	return int(codes.OK)
+}
+
+//export ClientRecvText
+func ClientRecvText(ctxC unsafe.Pointer, clientPtr uintptr, codeC *C.char, cb C.callback) int {
+	client, err := getClient(clientPtr)
+	if err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+	ctx := context.Background()
+
+	go func() {
+		msg, err := client.Receive(ctx, C.GoString(codeC), false)
+
+		if err != nil {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_RECV_TEXT))
+		}
+
+		data, err := ioutil.ReadAll(msg)
+		if err != nil {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_RECV_TEXT_DATA))
+		}
+		dataC := C.CBytes(data)
+		fileC := (*C.file_t)(C.malloc(C.sizeof_file_t))
```

Comment:
Was trying to understand how the callback works. So, the calling program to this library (dart or typescript or whatever) would be passing the callback function. And here we are filling in some memory and sending it to the callback.

I wonder how the allocated memory would get freed. If we are not freeing it, I guess if we use the program repeatedly without closing it, it would be leaking memory.